File: src/priv/bajdcc/LALR1/grammar/symbol/ManageScopeSymbol.java
package priv.bajdcc.LALR1.grammar.symbol;

import priv.bajdcc.LALR1.grammar.runtime.RuntimeObject;
import priv.bajdcc.LALR1.grammar.semantic.ISemanticRecorder;
import priv.bajdcc.LALR1.grammar.tree.Function;
import priv.bajdcc.LALR1.grammar.type.TokenTools;
import priv.bajdcc.util.HashListMap;
import priv.bajdcc.util.HashListMapEx;
import priv.bajdcc.util.Position;
import priv.bajdcc.util.lexer.token.Token;
import priv.bajdcc.util.lexer.token.TokenType;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Stack;

/**
 * 命名空间管理
 *
 * @author bajdcc
 */
public class ManageScopeSymbol implements IQueryScopeSymbol, IQueryBlockSymbol,
		IManageDataSymbol, IManageScopeSymbol {

	private static String ENTRY_NAME = "main";
	private static String LAMBDA_PREFIX = "~lambda#";
	private int lambdaId = 0;
	private HashListMap<Object> symbolList = new HashListMap<>();
	private HashListMapEx<String, Function> funcMap = new HashListMapEx<>();
	private ArrayList<HashSet<String>> stkScope = new ArrayList<>();
	private Stack<Integer> stkLambdaId = new Stack<>();
	private Stack<Integer> stkLambdaLine = new Stack<>();
	private HashSet<String> symbolsInFutureBlock = new HashSet<>();
	private HashMap<BlockType, Integer> blockLevel = new HashMap<>();
	private Stack<BlockType> blockStack = new Stack<>();

	public ManageScopeSymbol() {
		enterScope();
		funcMap.add(ENTRY_NAME, new Function());
		for (BlockType type : BlockType.values()) {
			blockLevel.put(type, 0);
		}
	}

	@Override
	public void enterScope() {
		stkScope.add(0, new HashSet<>());
		symbolsInFutureBlock.forEach(this::registerSymbol);
		clearFutureArgs();
	}

	@Override
	public void leaveScope() {
		stkScope.remove(0);
		clearFutureArgs();
	}

	@Override
	public void clearFutureArgs() {
		symbolsInFutureBlock.clear();
	}

	@Override
	public boolean findDeclaredSymbol(String name) {
		if (symbolsInFutureBlock.contains(name)) {
			return true;
		}
		for (HashSet<String> hashSet : stkScope) {
			if (hashSet.contains(name)) {
				return true;
			}
		}
		if (TokenTools.isExternalName(name)) {
			registerSymbol(name);
			return true;
		}
		return false;
	}

	@Override
	public boolean findDeclaredSymbolDirect(String name) {
		return symbolsInFutureBlock.contains(name) || stkScope.get(0).contains(name);
	}

	@Override
	public boolean isUniqueSymbolOfBlock(String name) {
		return stkScope.get(0).contains(name);
	}

	@Override
	public String getEntryName() {
		return ENTRY_NAME;
	}

	@Override
	public Token getEntryToken() {
		Token token = new Token();
		token.kToken = TokenType.ID;
		token.object = getEntryName();
		token.position = new Position();
		return token;
	}

	@Override
	public Function getFuncByName(String name) {
		Function func = funcMap.get(name);
		if (func != null) {
			return func;
		}
		int count = 0;
		for (Function function : funcMap.list) {
			String funcName = function.getRealName();
			if (funcName != null && funcName.equals(name)) {
				func = function;
				count++;
			}
		}
		// count=1 即无歧义, count>1 有歧义
		return (count == 1) ? func : null;
	}

	@Override
	public Function getFuncByRealName(String name) {
		for (Function func : funcMap.list) {
			if (name.equals(func.getRealName())) {
				return func;
			}
		}
		return null;
	}

	@Override
	public Function getLambda() {
		int lambdaId = stkLambdaId.pop();
		int lambdaLine = stkLambdaLine.pop();
		return funcMap.get(LAMBDA_PREFIX + lambdaId + "!" + lambdaLine);
	}

	@Override
	public void registerSymbol(String name) {
		stkScope.get(0).add(name);
		symbolList.add(name);
	}

	@Override
	public void registerFunc(Function func) {
		if (func.getName().kToken == TokenType.ID) {
			func.setRealName(func.getName().toRealString());
			symbolList.add(func.getRealName());
		} else {
			func.setRealName(LAMBDA_PREFIX + lambdaId++);
		}
		funcMap.add(func.getRealName(), func);
	}

	@Override
	public void registerLambda(Function func) {
		stkLambdaId.push(lambdaId);
		stkLambdaLine.push(func.getName().position.iLine);
		func.getName().kToken = TokenType.ID;
		func.setRealName(LAMBDA_PREFIX + (lambdaId++) + "!" + stkLambdaLine.peek());
		funcMap.add(func.getRealName(), func);
	}

	@Override
	public boolean isRegisteredFunc(String name) {
		Function func = funcMap.get(name);
		return func != null && func.getRealName() != null;
	}

	@Override
	public boolean registerFutureSymbol(String name) {
		return symbolsInFutureBlock.add(name);
	}

	public void check(ISemanticRecorder recorder) {
		for (Function func : funcMap.list) {
			func.analysis(recorder);
		}
	}

	@Override
	public HashListMap<Object> getSymbolList() {
		return symbolList;
	}

	@Override
	public HashListMapEx<String, Function> getFuncMap() {
		return funcMap;
	}

	public String getSymbolString() {
		StringBuilder sb = new StringBuilder();
		sb.append("#### 符号表 ####");
		sb.append(System.lineSeparator());
		int i = 0;
		for (Object symbol : symbolList.list) {
			sb.append(i).append(": ").append("[").append(RuntimeObject.fromObject(symbol).getName()).append("] ").append(symbol);
			sb.append(System.lineSeparator());
			i++;
		}
		return sb.toString();
	}

	public String getFuncString() {
		StringBuilder sb = new StringBuilder();
		sb.append("#### 过程表 ####");
		sb.append(System.lineSeparator());
		int i = 0;
		for (Function func : funcMap.list) {
			sb.append("----==== #").append(i).append(" ====----");
			sb.append(System.lineSeparator());
			sb.append(func.toString());
			sb.append(System.lineSeparator());
			sb.append(System.lineSeparator());
			i++;
		}
		return sb.toString();
	}

	@Override
	public String toString() {
		String sb = getSymbolString() +
				getFuncString();
		return sb;
	}

	@Override
	public void enterBlock(BlockType type) {
		switch (type) {
			case kCycle:
				int level = blockLevel.get(type);
				blockLevel.put(type, level + 1);
				break;
			case kFunc:
			case kYield:
				blockStack.push(type);
				break;
			default:
				break;

		}
	}

	@Override
	public void leaveBlock(BlockType type) {
		switch (type) {
			case kCycle:
				int level = blockLevel.get(type);
				blockLevel.put(type, level - 1);
				break;
			case kFunc:
			case kYield:
				if (blockStack.peek() == type) {
					blockStack.pop();
				}
				break;
			default:
				break;
		}
	}

	@Override
	public boolean isInBlock(BlockType type) {
		switch (type) {
			case kCycle:
				return blockLevel.get(type) > 0;
			case kFunc:
			case kYield:
				return !blockStack.isEmpty() && (blockStack.peek() == type);
			default:
				break;
		}
		return false;
	}
}

File: src/main/java/com/bajdcc/LALR1/grammar/symbol/ManageScopeSymbol.kt
package com.bajdcc.LALR1.grammar.symbol

import com.bajdcc.LALR1.grammar.runtime.RuntimeObject
import com.bajdcc.LALR1.grammar.semantic.ISemanticRecorder
import com.bajdcc.LALR1.grammar.tree.Function
import com.bajdcc.LALR1.grammar.type.TokenTools
import com.bajdcc.util.HashListMap
import com.bajdcc.util.HashListMapEx
import com.bajdcc.util.Position
import com.bajdcc.util.lexer.token.Token
import com.bajdcc.util.lexer.token.TokenType
import java.util.*

/**
 * 命名空间管理
 *
 * @author bajdcc
 */
class ManageScopeSymbol : IQueryScopeSymbol, IQueryBlockSymbol, IManageDataSymbol, IManageScopeSymbol {
    private var lambdaId = 0
    override val symbolList = HashListMap<Any>()
    override val funcMap = HashListMapEx<String, MutableList<Function>>()
    private val funcScope = mutableListOf<MutableMap<String, Function>>()
    private val stkScope = mutableListOf<MutableSet<String>>()
    private val stkLambdaId = Stack<Int>()
    private val stkLambdaLine = Stack<Int>()
    private val symbolsInFutureBlock = mutableSetOf<String>()
    private val blockLevel = mutableMapOf<BlockType, Int>()
    private val blockStack = Stack<BlockType>()

    override val entryName: String
        get() = ENTRY_NAME

    override val entryToken: Token
        get() {
            val token = Token()
            token.type = TokenType.ID
            token.obj = entryName
            token.position = Position()
            return token
        }

    override val lambda: Function
        get() {
            val lambdaId = stkLambdaId.pop()
            val lambdaLine = stkLambdaLine.pop()
            return funcMap["$LAMBDA_PREFIX$lambdaId!$lambdaLine"][0]
        }

    private val symbolString: String
        get() {
            val sb = StringBuilder()
            sb.append("#### 符号表 ####")
            sb.append(System.lineSeparator())
            symbolList.toList().withIndex().forEach { (i, symbol) ->
                sb.append(i).append(": ").append("[").append(RuntimeObject.fromObject(symbol).desc).append("] ").append(symbol)
                sb.append(System.lineSeparator())
            }
            return sb.toString()
        }

    private val funcString: String
        get() {
            val sb = StringBuilder()
            sb.append("#### 过程表 ####")
            sb.append(System.lineSeparator())
            var i = 0
            for (funcs in funcMap.toList()) {
                for (func in funcs) {
                    sb.append("----==== #").append(i).append(" ====----")
                    sb.append(System.lineSeparator())
                    sb.append(func.toString())
                    sb.append(System.lineSeparator())
                    sb.append(System.lineSeparator())
                    i++
                }
            }
            return sb.toString()
        }

    init {
        enterScope()
        val entry = mutableListOf<Function>()
        entry.add(Function(Token()))
        funcMap.add(ENTRY_NAME, entry)
        for (type in BlockType.values()) {
            blockLevel[type] = 0
        }
    }

    private val currentScope: MutableSet<String>
        get() = stkScope[stkScope.size - 1]

    private val currentFuncScope: MutableMap<String, Function>
        get() = funcScope[funcScope.size - 1]

    override fun enterScope() {
        stkScope.add(mutableSetOf())
        funcScope.add(mutableMapOf())
        symbolsInFutureBlock.forEach { this.registerSymbol(it) }
        clearFutureArgs()
    }

    override fun leaveScope() {
        stkScope.removeAt(stkScope.size - 1)
        funcScope.removeAt(funcScope.size - 1)
        clearFutureArgs()
    }

    override fun clearFutureArgs() {
        symbolsInFutureBlock.clear()
    }

    override fun findDeclaredSymbol(name: String): Boolean {
        if (symbolsInFutureBlock.contains(name)) {
            return true
        }
        if (stkScope.reversed().any { it.contains(name) })
            return true
        if (TokenTools.isExternalName(name)) {
            registerSymbol(name)
            return true
        }
        return false
    }

    override fun findDeclaredSymbolDirect(name: String): Boolean {
        return symbolsInFutureBlock.contains(name) || currentScope.contains(name)
    }

    override fun isUniqueSymbolOfBlock(name: String): Boolean {
        return currentScope.contains(name)
    }

    override fun getFuncByName(name: String): Function? {
        funcScope.indices.reversed().forEach { i ->
            val f = funcScope[i]
            val f1 = f[name]
            if (f1 != null)
                return f1
            for (func in f.values) {
                val funcName = func.realName
                if (funcName == name) {
                    return func
                }
            }
        }
        return null
    }

    override fun isLambda(name: String): Boolean {
        return name.startsWith(LAMBDA_PREFIX)
    }

    override fun registerSymbol(name: String) {
        currentScope.add(name)
        symbolList.add(name)
    }

    override fun registerFunc(func: Function) {
        if (func.name.type === TokenType.ID) {
            func.realName = func.name.toRealString()
            symbolList.add(func.realName)
        } else {
            func.realName = LAMBDA_PREFIX + lambdaId++
        }
        val f = mutableListOf<Function>()
        f.add(func)
        funcMap.add(func.realName, f)
        currentFuncScope[func.realName] = func
    }

    override fun registerLambda(func: Function) {
        stkLambdaId.push(lambdaId)
        stkLambdaLine.push(func.name.position.line)
        func.name.type = TokenType.ID
        func.realName = LAMBDA_PREFIX + lambdaId++ + "!" + stkLambdaLine.peek()
        currentFuncScope[func.realName] = func
        val f = mutableListOf<Function>()
        f.add(func)
        funcMap.add(func.realName, f)
    }

    override fun isRegisteredFunc(name: String): Boolean {
        return funcMap.contains(name)
    }

    override fun registerFutureSymbol(name: String): Boolean {
        return symbolsInFutureBlock.add(name)
    }

    fun check(recorder: ISemanticRecorder) {
        funcMap.toList().flatMap { it }.forEach { it.analysis(recorder) }
    }

    override fun toString(): String {
        return symbolString + funcString
    }

    override fun enterBlock(type: BlockType) {
        when (type) {
            BlockType.kCycle -> {
                val level = blockLevel[type]!!
                blockLevel[type] = level + 1
            }
            BlockType.kFunc, BlockType.kYield -> blockStack.push(type)
        }
    }

    override fun leaveBlock(type: BlockType) {
        when (type) {
            BlockType.kCycle -> {
                val level = blockLevel[type]!!
                blockLevel[type] = level - 1
            }
            BlockType.kFunc, BlockType.kYield -> if (blockStack.peek() == type) {
                blockStack.pop()
            }
        }
    }

    override fun isInBlock(type: BlockType): Boolean {
        return when (type) {
            BlockType.kCycle -> blockLevel[type]!! > 0
            BlockType.kFunc, BlockType.kYield -> !blockStack.isEmpty() && blockStack.peek() == type
        }
    }

    companion object {
        private const val ENTRY_NAME = "main"
        private const val LAMBDA_PREFIX = "~lambda#"
    }
}
